### ISSUE TITLE
Datapoint update

### ISSUE THREAD

**Issue by nick-demianchuk** [VictoriaMetrics/VictoriaMetrics#844]
**Is your feature request related to a problem? Please describe.**
Our company's use case is to be able to update a value for specific metric and specific timestamp. 
Metrics are not being collected in real time, but rather computed based on the events flow, and considering that some events might be late - metric-generating jobs might send an updated value for a specific bucket. 
Currently company is using Graphite as a data storage, and it allows updates within a selected retention, but it would be nice to have something similar in Victoria Metrics.

**Describe the solution you'd like**
Option for a metric with the same name, fields set and timestamp to update previously sent value. 
Preferably, any datapoint should be updatable, even the most old one.

**Comment by valyala** [comment]
Unfortunately it can be hard implementing datapoint value updates in VictoriaMetrics due to the chosen data storage design - VictoriaMetrics stores all the datapoints with identical timestamps under the given time series.

There is a workaround, which allows "updating" datapoint value with specific timestamps - to store delta values with identical timestamps and to use `sum_over_time()` function for obtaining actual value. For example, if there is a datapoint with value `123` and it should be updated to value `42`, then just store delta=`42 - 123 = -81` with the same timestamp as the original value `123`. Now `sum_over_time()` for this time series will return `123 + (-81) = 42`.

Let's keep this feature request open in the hope better solution will appear.

**Comment by Roemer** [comment]
ChatGPT told me that I can just send a new datapoint with the same labels and timestamp and the new value will overwrite the old value, is that correct as of now?

Quote:
> For example, let's say you have a data point for metric cpu_usage with a value of 50 and a timestamp of 1620352830. If you later discover that the correct value for that data point should have been 60, you can send a new data point with the same timestamp and labels, but with the corrected value of 60:
> 
> ```bash
> POST /api/v1/write HTTP/1.1
> Host: <victoriametrics-hostname>:8428
> cpu_usage{label1="value1", label2="value2"} 60 1620352830
> ```
> After this data point is ingested, the previous data point with a value of 50 will be replaced by the new data point with a value of 60. When you query the cpu_usage metric for the specific label values, you will get the corrected value of 60 instead of the original value of 50.

**Comment by zekker6** [comment]
@Roemer That's not correct, both data points will be stored in the database.
Although, you can use [deduplication](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#deduplication) to achieve update-like behavior for your data. You can write new data point 1 ms after the previous one was set and enable `-dedup.minScrapeInterval=10ms`(note that it should be present at vmselect and vmstorage for cluster version). In this case when query will be processed and database will identify 2 data points within deduplication interval it will pick value with the highest timestamp. Using `10ms` interval gives you an option to override datapoint multiple times.
Note that using this way will have a performance penalty since deduplication will now check for duplicates every time query is processed.

**Comment by hagen1778** [comment]
related issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3992

**Comment by namirsab** [comment]
@zekker6  In your last comment you mention this: 
> Using `10ms` interval gives you an option to override datapoint multiple times. Note that using this way will have a performance penalty since deduplication will now check for duplicates every time query is processed.

How many times exactly? I understand this would be 10 times - increasing the timestamp 1 ms every time we get a new value for the same timestamp -, or can we also use fractions (e.g. nanoseconds)?

We are analysing different ways to circumvent this issue, and this one sounds like the most promising

Thanks for your help! :)

**Comment by zekker6** [comment]
@namirsab 
> How many times exactly? I understand this would be 10 times - increasing the timestamp 1 ms every time we get a new value for the same timestamp -, or can we also use fractions (e.g. nanoseconds)?

That's correct, that will be 10 times. VictoriaMetrics stores timestamps with millisecond precision, so you can't use smaller fractions.

**Comment by namirsab** [comment]
> That's correct, that will be 10 times. VictoriaMetrics stores timestamps with millisecond precision, so you can't use smaller fractions.

Thanks a lot @zekker6 

Another question we have is when is the bucket starting:
is it from 10:00 to 10:09 or from 10:01 to 10:10?

I hope the question is clear :)

**Comment by zekker6** [comment]
@namirsab It will be `10:00 to 10:09`.

**Comment by namirsab** [comment]
Okay, it seems I just randomly hit the end of the 10ms window. I added a third value and now it only shows the first one and the last one I've added. 
I think because the first one is in a different window than the other 2

**Comment by namirsab** [comment]
After some testing, I think the window is actually going from 
1685976488001 until 1685976488010 and not from 1685976488000 until 1685976488009 as I expected.

This explains the behaviour I had yesterday

**Comment by valyala** [comment]
We aren't going to implement update API, since VictoriaMetrics architecture wasn't designed with the update API in mind. There was an unsuccessful attempt to add support for update API - see https://github.com/VictoriaMetrics/VictoriaMetrics/pull/2885 . This attempt was used by one of our customers in production, but it had performance and resource usage issues, so we decided to abandon it.

The basic idea was to store a `shadow` time series with every `update` call, and then to merge all the `shadow` time series with the original time series at query time. This didn't work well when the `update` was called a lot of times. This resulted in millions of `shadow` time series, which were occupying significant amounts of RAM and CPU during querying.

Closing this feature request as won't implement.
